diff --git a/libavformat/Makefile b/libavformat/Makefile
index 7ca68a7036..eb9bc51321 100644
--- a/libavformat/Makefile
+++ b/libavformat/Makefile
@@ -4,6 +4,14 @@ DESC = FFmpeg container format library
 HEADERS = avformat.h                                                    \
           avio.h                                                        \
           version.h                                                     \
+          avc.h                                                         \
+          url.h                                                         \
+          internal.h                                                    \
+          avio_internal.h                                               \
+          flv.h                                                         \
+          id3v2.h                                                       \
+          os_support.h                                                  \
+          metadata.h                                                  \
           version_major.h                                               \
 
 OBJS = allformats.o         \
@@ -29,6 +37,8 @@ OBJS = allformats.o         \
        seek.o               \
        url.o                \
        utils.o              \
+       avc.o                \
+       ijkutils.o           \
        version.o            \
 
 OBJS-$(HAVE_LIBC_MSVCRT)                 += file_open.o