diff --git a/libavformat/allformats.c b/libavformat/allformats.c
index 305fa46532..e9a2de7613 100644
--- a/libavformat/allformats.c
+++ b/libavformat/allformats.c
@@ -531,6 +531,8 @@ extern const FFInputFormat  ff_xwma_demuxer;
 extern const FFInputFormat  ff_yop_demuxer;
 extern const FFInputFormat  ff_yuv4mpegpipe_demuxer;
 extern const FFOutputFormat ff_yuv4mpegpipe_muxer;
+extern const FFInputFormat  ff_ijklivehook_demuxer;
+extern const FFInputFormat  ff_ijklas_demuxer;
 /* image demuxers */
 extern const FFInputFormat  ff_image_bmp_pipe_demuxer;
 extern const FFInputFormat  ff_image_cri_pipe_demuxer;
