diff --git a/libavformat/protocols.c b/libavformat/protocols.c
index 93a6d67261..6313635d53 100644
--- a/libavformat/protocols.c
+++ b/libavformat/protocols.c
@@ -44,6 +44,12 @@ extern const URLProtocol ff_http_protocol;
 extern const URLProtocol ff_httpproxy_protocol;
 extern const URLProtocol ff_https_protocol;
 extern const URLProtocol ff_icecast_protocol;
+extern const URLProtocol ff_ijkhttphook_protocol;
+extern const URLProtocol ff_ijklongurl_protocol;
+extern const URLProtocol ff_ijkmediadatasource_protocol;
+extern const URLProtocol ff_ijksegment_protocol;
+extern const URLProtocol ff_ijktcphook_protocol;
+extern const URLProtocol ff_ijkio_protocol;
 extern const URLProtocol ff_mmsh_protocol;
 extern const URLProtocol ff_mmst_protocol;
 extern const URLProtocol ff_md5_protocol;