diff --git a/libavformat/concatdec.c b/libavformat/concatdec.c
index fe65d0c768..d73f4f238f 100644
--- a/libavformat/concatdec.c
+++ b/libavformat/concatdec.c
@@ -72,6 +72,8 @@ typedef struct {
     ConcatMatchMode stream_match_mode;
     unsigned auto_convert;
     int segment_time_metadata;
+    AVDictionary *options;
+    int error;
 } ConcatContext;
 
 static int concat_probe(const AVProbeData *probe)
@@ -336,38 +338,61 @@ static int open_file(AVFormatContext *avf, unsigned fileno)
 {
     ConcatContext *cat = avf->priv_data;
     ConcatFile *file = &cat->files[fileno];
-    AVDictionary *options = NULL;
+    AVFormatContext *new_avf = NULL;
     int ret;
+    AVDictionary *tmp = NULL;
+    AVDictionaryEntry *t = NULL;
+    int fps_flag = 0;
 
-    if (cat->avf)
-        avformat_close_input(&cat->avf);
-
-    cat->avf = avformat_alloc_context();
-    if (!cat->avf)
+    new_avf = avformat_alloc_context();
+    if (!new_avf)
         return AVERROR(ENOMEM);
 
-    cat->avf->flags |= avf->flags & ~AVFMT_FLAG_CUSTOM_IO;
-    cat->avf->interrupt_callback = avf->interrupt_callback;
+    new_avf->flags |= avf->flags & ~AVFMT_FLAG_CUSTOM_IO;
+    new_avf->interrupt_callback = avf->interrupt_callback;
 
-    if ((ret = ff_copy_whiteblacklists(cat->avf, avf)) < 0)
+    if ((ret = ff_copy_whiteblacklists(new_avf, avf)) < 0)
         return ret;
 
-    ret = av_dict_copy(&options, file->options, 0);
+    if (cat->options)
+        av_dict_copy(&tmp, cat->options, 0);
+
+    av_dict_set_int(&tmp, "cur_file_no", fileno, 0);
+
+    t = av_dict_get(tmp, "skip-calc-frame-rate", NULL, AV_DICT_MATCH_CASE);
+    if (t) {
+        fps_flag = (int) strtol(t->value, NULL, 10);
+        if (fps_flag > 0)
+            av_dict_set_int(&new_avf->metadata, "skip-calc-frame-rate", fps_flag, 0);
+    }
+
+    t = av_dict_get(tmp, "nb-streams", NULL, AV_DICT_MATCH_CASE);
+    if (t) {
+        int nb_streams = (int) strtol(t->value, NULL, 10);
+        if (nb_streams > 0) {
+            av_dict_set_int(&new_avf->metadata, "nb-streams", nb_streams, 0);
+            av_dict_set_int(&cat->options, "nb-streams", 0, 0);
+        }
+    }
+
+    ret = av_dict_copy(&tmp, file->options, 0);
     if (ret < 0)
         return ret;
 
-    if ((ret = avformat_open_input(&cat->avf, file->url, NULL, &options)) < 0 ||
-        (ret = avformat_find_stream_info(cat->avf, NULL)) < 0) {
-        av_log(avf, AV_LOG_ERROR, "Impossible to open '%s'\n", file->url);
-        av_dict_free(&options);
-        avformat_close_input(&cat->avf);
+    ret = avformat_open_input(&new_avf, file->url, NULL, &tmp);
+    av_dict_free(&tmp);
+    if (ret < 0 ||
+        (ret = avformat_find_stream_info(new_avf, NULL)) < 0) {
+        av_log(avf, AV_LOG_ERROR, "Impossible to open '%s\n", file->url);
+        avformat_close_input(&new_avf);
         return ret;
     }
-    if (options) {
-        av_log(avf, AV_LOG_WARNING, "Unused options for '%s'.\n", file->url);
-        /* TODO log unused options once we have a proper string API */
-        av_dict_free(&options);
-    }
+
+    if (cat->avf)
+        avformat_close_input(&cat->avf);
+
+    avf->bit_rate = new_avf->bit_rate;
+    cat->avf      = new_avf;
     cat->cur_file = file;
     file->start_time = !fileno ? 0 :
                        cat->files[fileno - 1].start_time +
@@ -408,6 +433,7 @@ static int concat_read_close(AVFormatContext *avf)
     }
     if (cat->avf)
         avformat_close_input(&cat->avf);
+    av_dict_free(&cat->options);
     av_freep(&cat->files);
     return 0;
 }
@@ -657,13 +683,16 @@ fail:
     return ret == AVERROR_EOF ? 0 : ret;
 }
 
-static int concat_read_header(AVFormatContext *avf)
+static int concat_read_header2(AVFormatContext *avf, AVDictionary **options)
 {
     ConcatContext *cat = avf->priv_data;
     int64_t time = 0;
     unsigned i;
     int ret;
 
+    if (options && *options)
+        av_dict_copy(&cat->options, *options, 0);
+
     ret = concat_parse_script(avf);
     if (ret < 0)
         return ret;
@@ -750,6 +779,8 @@ static int packet_after_outpoint(ConcatContext *cat, AVPacket *pkt)
     return 0;
 }
 
+#define CONCAT_MAX_OPEN_TRY 3
+
 static int concat_read_packet(AVFormatContext *avf, AVPacket *pkt)
 {
     ConcatContext *cat = avf->priv_data;
@@ -758,6 +789,11 @@ static int concat_read_packet(AVFormatContext *avf, AVPacket *pkt)
     ConcatStream *cs;
     AVStream *st;
     FFStream *sti;
+    int try_counter = 0;
+    int is_new_st = 0;
+
+    if (cat->error)
+        return cat->error;
 
     if (cat->eof)
         return AVERROR_EOF;
@@ -768,19 +804,27 @@ static int concat_read_packet(AVFormatContext *avf, AVPacket *pkt)
     while (1) {
         ret = av_read_frame(cat->avf, pkt);
         if (ret == AVERROR_EOF) {
+            is_new_st = 1;
             if ((ret = open_next_file(avf)) < 0)
-                return ret;
+                goto open_fail;
             continue;
         }
-        if (ret < 0)
+        if (ret < 0) {
+            if (avf->pb && cat->avf->pb)
+                avf->pb->error = cat->avf->pb->error;
             return ret;
+        }
+        if (is_new_st) {
+            pkt->flags |= AV_PKT_FLAG_NEW_SEG;
+            is_new_st = 0;
+        }
         if ((ret = match_streams(avf)) < 0) {
             return ret;
         }
         if (packet_after_outpoint(cat, pkt)) {
             av_packet_unref(pkt);
             if ((ret = open_next_file(avf)) < 0)
-                return ret;
+                goto open_fail;
             continue;
         }
         cs = &cat->cur_file->streams[pkt->stream_index];
@@ -789,6 +833,15 @@ static int concat_read_packet(AVFormatContext *avf, AVPacket *pkt)
             continue;
         }
         break;
+open_fail:
+        ++try_counter;
+        if (try_counter > CONCAT_MAX_OPEN_TRY) {
+            cat->error = ret;
+            if (avf->pb && ret != AVERROR_EOF)
+                avf->pb->error = ret;
+            return AVERROR_EOF;
+        }
+        av_log(avf, AV_LOG_WARNING, "open_next_file() failed (%d)\n", try_counter);
     }
     if ((ret = filter_packet(avf, cs, pkt)) < 0)
         return ret;
@@ -910,6 +963,9 @@ static int concat_seek(AVFormatContext *avf, int stream,
     AVFormatContext *cur_avf_saved = cat->avf;
     int ret;
 
+    /* reset error/complete state */
+    cat->error = 0;
+
     if (flags & (AVSEEK_FLAG_BYTE | AVSEEK_FLAG_FRAME))
         return AVERROR(ENOSYS);
     cat->avf = NULL;
@@ -957,7 +1013,7 @@ const FFInputFormat ff_concat_demuxer = {
     .priv_data_size = sizeof(ConcatContext),
     .flags_internal = FF_INFMT_FLAG_INIT_CLEANUP,
     .read_probe     = concat_probe,
-    .read_header    = concat_read_header,
+    .read_header2   = concat_read_header2,
     .read_packet    = concat_read_packet,
     .read_close     = concat_read_close,
     .read_seek2     = concat_seek,