diff --git a/libavformat/network.h b/libavformat/network.h
index ca214087fc..696762a7a5 100644
--- a/libavformat/network.h
+++ b/libavformat/network.h
@@ -304,6 +304,11 @@ int ff_listen_connect(int fd, const struct sockaddr *addr,
                       socklen_t addrlen, int timeout,
                       URLContext *h, int will_try_next);
 
+int ff_sendto(int fd, const char *msg, int msg_len, int flag,
+                      const struct sockaddr *addr,
+                      socklen_t addrlen, int timeout, URLContext *h,
+                      int will_try_next);
+
 int ff_http_match_no_proxy(const char *no_proxy, const char *hostname);
 
 int ff_socket(int domain, int type, int protocol, void *logctx);