diff --git a/libavutil/error.h b/libavutil/error.h
index 1efa86c4c1..9dee755c69 100644
--- a/libavutil/error.h
+++ b/libavutil/error.h
@@ -85,6 +85,10 @@
 
 #define AV_ERROR_MAX_STRING_SIZE 64
 
+#define AVERROR_TCP_CONNECT_TIMEOUT -1001
+#define AVERROR_TCP_READ_TIMEOUT -1002
+#define AVERROR_TCP_WRITE_TIMEOUT -1003
+
 /**
  * Put a description of the AVERROR code errnum in errbuf.
  * In case of failure the global variable errno is set to indicate the
