diff --git a/libavformat/async.c b/libavformat/async.c
index e0329e23ec..e5ee519354 100644
--- a/libavformat/async.c
+++ b/libavformat/async.c
@@ -489,7 +489,7 @@ static const AVClass async_context_class = {
     .version    = LIBAVUTIL_VERSION_INT,
 };
 
-const URLProtocol ff_async_protocol = {
+URLProtocol ff_async_protocol = {
     .name                = "async",
     .url_open2           = async_open,
     .url_read            = async_read,
