diff --git a/libavformat/avio.c b/libavformat/avio.c
index d109f3adff..8d72dbe1a8 100644
--- a/libavformat/avio.c
+++ b/libavformat/avio.c
@@ -532,10 +532,8 @@ static inline int retry_transfer_wrapper(URLContext *h, uint8_t *buf,
                 }
                 av_usleep(1000);
             }
-        } else if (ret == AVERROR_EOF)
-            return (len > 0) ? len : AVERROR_EOF;
-        else if (ret < 0)
-            return ret;
+        } else if (ret < 1)
+            return (ret < 0 && ret != AVERROR_EOF) ? ret : len;
         if (ret) {
             fast_retries = FFMAX(fast_retries, 2);
             wait_since = 0;
