diff --git a/libavformat/concat.c b/libavformat/concat.c
index e1d57de557..b6bb99f2f3 100644
--- a/libavformat/concat.c
+++ b/libavformat/concat.c
@@ -142,20 +142,19 @@ static int concat_read(URLContext *h, unsigned char *buf, int size)
 
     while (size > 0) {
         result = ffurl_read(nodes[i].uc, buf, size);
-        if (result == AVERROR_EOF) {
+        if (result < 0)
+            return total ? total : result;
+        if (!result) {
             if (i + 1 == data->length ||
                 ffurl_seek(nodes[++i].uc, 0, SEEK_SET) < 0)
                 break;
-            result = 0;
         }
-        if (result < 0)
-            return total ? total : result;
         total += result;
         buf   += result;
         size  -= result;
     }
     data->current = i;
-    return total ? total : result;
+    return total;
 }
 
 static int64_t concat_seek(URLContext *h, int64_t pos, int whence)