diff --git a/libavformat/isom.h b/libavformat/isom.h
index 4723397048..80e57239fd 100644
--- a/libavformat/isom.h
+++ b/libavformat/isom.h
@@ -349,6 +349,8 @@ typedef struct MOVContext {
     uint8_t *decryption_key;
     int decryption_key_len;
     int enable_drefs;
+    int allow_multi_extradata;
+    int has_extradata;
     int32_t movie_display_matrix[3][3]; ///< display matrix from mvhd
     int have_read_mfra_size;
     uint32_t mfra_size;
